### ISSUE TITLE
Skip to the next trial

### ISSUE THREAD

**Issue by acegilz** [keras-team/keras-tuner#279]
Is there any possibly skip to the next epochs in case a trial is stalled, or starts decreasing the goal? (defined under certain custom options)

Also, I think there should be a way of training the model (model.fit) with these "random" hyperparams, instead of just determining a good set of hyperparams to train later.

I know this may be another concept, but I will find it very useful. 

For example to train a model during the whole night with the same params, in my case after 1 hour (several epochs) the model is already wellll overtrained. If I could be training multiple models with random params automatically to try find a better accuracy it would be awesome.

Consider this please..

**Comment by ben-arnao** [comment]
Unless i'm misunderstanding this isn't really a question to do with keras-tuner but rather training in general.

`Is there any possibly skip to the next epochs in case a trial is stalled, or starts decreasing the goal`

If you mean that you want to stop training the current model and start a new trial when loss stopd decreasing/starting over fitting and going up there is a callback for that

https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/EarlyStopping

`Also, I think there should be a way of training the model (model.fit) with these "random" hyperparams, instead of just determining a good set of hyperparams to train later.`

Not really sure what you are asking here but RandomSearch may be what you want. You get a random set of hyper parameters to build and train your model from. Although i'm not sure why you would prefer this over BayesianOptimization.

**Comment by acegilz** [comment]
Hello,

Yes I am using Random Search but my idea would be to go to the next trial skipping missing epochs, if the rate starts decreasing. is it possible to implement it with the link you sent for the keras tuner?

**Comment by ben-arnao** [comment]
> Hello,
> 
> Yes I am using Random Search but my idea would be to go to the next trial skipping missing epochs, if the rate starts decreasing. is it possible to implement it with the link you sent for the keras tuner?
> 
> I haven't tried Bayesian yet but the idea would be to automatically test several params with different models, instead of no trains at all (currently it doesn't save trains/models, just the params afaik)

Yes, keras-tuner supports fit() arguments like `callbacks` directly in search()

```
tuner.search(train_ds,
             validation_data=test_ds,
             callbacks=[tf.keras.callbacks.EarlyStopping(patience=1)])
```

But more generally kt supports any methodology you want to use to assign score to a trial

```
class MyTuner(kt.Tuner):

    def run_trial(self, trial, ...):
        model = self.hypermodel.build(trial.hyperparameters)
        score = ...
        self.oracle.update_trial(trial.trial_id, {'score': score})
```

So that includes callbacks or any custom train logic you want to use.

**Comment by acegilz** [comment]
this was very helpful Ben,
thank you!

**Comment by acegilz** [comment]
I am using the EarlyStopping callback you recommended above.

Any way of "deactivating" while after a certain point? Example: While the val_accuracy is above 60%, it should keep testing and do not Stop. 

I have tried with _baseline_:

```python
tf.keras.callbacks.EarlyStopping(
        monitor='val_accuracy',
        baseline=0.6,
        patience=15)
```

but it does not work. Any way of achieving this? I just don't want to stop it if the val_Accuracy is below 60%, even if it's not growing.

**Comment by ben-arnao** [comment]
@acegilz Again this is more of a general/tensorflow question than a keras-tuner one, but the source code for these callbacks is available and very simple to replicate to use to your own specifications.

https://github.com/tensorflow/tensorflow/blob/v2.1.0/tensorflow/python/keras/callbacks.py#L1170-L1287

If the standard callback isn't providing the functionality you need you could just hard code a line to the epoch_end method to suit your needs. If you think this is something you'll be doing a lot just rewrite the callback and add an extra parameter instead.

```
def on_epoch_end(self, epoch, logs=None):
    current = self.get_monitor_value(logs)
    if current is None:
      return
    # do not even execute the rest of the callback if the metric is above 0.6
    if current > 0.6:
      return
    ...
```
